src/Main.elm

```diff
@@ -1,10 +1,829 @@
 module Main exposing (main)
 
+import Browser
+import Browser.Events
 import Data.Audience
 import Data.AudienceFolder
-import Html exposing (Html)
+import Dict
+import Element as E -- elm-ui
+import Element.Background as EBackground
+import Element.Border as EBorder
+import Element.Events as EEvents
+import Element.Font as EFont
+import Element.Input as EInput
+import Html
+import Html.Attributes
+import Icons
+import Json.Decode
+import Palette
+import Process
+import Task
+import Time
 
 
-main : Html msg
+
+-- we could perhaps model this so that you can only make a "FolderClicked" Msg
+-- or "SortBy" Msg only after the model has loaded up, but I don't think this is
+-- too much of an issue
+
+
+type Msg
+    = FolderClicked Int
+    | GoBack
+    | SortBy Data.Audience.AudienceType
+    | GotAudiencesJSON String
+    | GotAudienceFoldersJSON String
+    | GotTime Time.Posix
+
+
+
+-- we use dicts instead of lists because it gives us O(log n) time complexity
+-- for insertion, removal and querying, as opposed to lists which would be O(n).
+-- additional note: I chose to store audiences and audienceFolders separately because:
+--    1. I don't know if on the server side the IDs of the Folders
+--       and the IDs of the Audiences are kept separately. This way I
+--       would make sure that there's no collisions between them
+--    2. These two come from different API endpoints.
+--       Audiences come from "/api/audiences" and audienceFolders come
+--       from "api/audience_folders".
+--       If we kept these as the same Dict, and only one result would come back,
+--       the dict would look completely normal, but it wouldn't be. Things would
+--       be missing.
+--       So this way, it's either "everything came back, and we render the ui,
+--       or we render some loading page or some error page "
+
+
+type Model
+    = Loading LoadingModel
+    | Loaded LoadedModel
+    | JsonError Json.Decode.Error
+
+
+type alias LoadingModel =
+    { audiences : Maybe (Dict.Dict Int FormattedAudience)
+    , audienceFolders : Maybe (Dict.Dict Int FormattedAudienceFolder)
+    , time : Maybe Time.Posix
+    }
+
+
+type alias LoadedModel =
+    { audiences : Dict.Dict Int FormattedAudience
+    , audienceFolders : Dict.Dict Int FormattedAudienceFolder
+
+    -- we store the full path, because we want to record the full history
+    -- to go back in the tree of folders
+    , currentFolderPath : List Int
+
+    -- according to "SECOND_STEP.md" there must be at least one category selected
+    -- at any time
+    , sortBy : Data.Audience.AudienceType
```

Comment:
In fact it's not sorting, but filtering.

Reply:
You're right, my mistake.